src/group_theory/commensurable.lean

```diff
@@ -0,0 +1,118 @@
+  commensurable H L :=
+⟨subgroup.relindex_ne_zero_trans H K L hhk.1 hkl.1,
+  subgroup.relindex_ne_zero_trans L K H hkl.2 hhk.2⟩
+
+lemma equivalence : equivalence (@commensurable G _) :=
+⟨commensurable.refl, λ _ _, commensurable.symm, λ _ _ _, commensurable.trans⟩
+
+/--Equivalence of `K/H ⊓ K` with `gKg⁻¹/gHg⁻¹ ⊓ gKg⁻¹`-/
+noncomputable def  quot_conj_equiv (H K : subgroup G) (g : G) :
+  quotient_group.quotient (H.subgroup_of K) ≃
+  quotient_group.quotient ((subgroup.conj_subgroup g H).subgroup_of (subgroup.conj_subgroup g K)) :=
+begin
+  have h1 := quotient_group.quot_map_inj (H.subgroup_of K) ((subgroup.conj_subgroup g H).subgroup_of
+    (subgroup.conj_subgroup g K))(subgroup.conj_equiv g K).to_monoid_hom
+    (subgroup.cong_sub_image H K g) (subgroup.conj_equiv g K).injective,
+  have h2 :=  quotient_group.quot_map_inj ((subgroup.conj_subgroup g H).subgroup_of
+    (subgroup.conj_subgroup g K)) (H.subgroup_of K) ((subgroup.conj_equiv g K).symm).to_monoid_hom
+    (subgroup.cong_sub_image' H K g) (subgroup.conj_equiv g K).symm.injective,
+  have := function.embedding.schroeder_bernstein  h1 h2,
+  apply equiv.of_bijective this.some,
+  apply this.some_spec,
+end
+
+lemma commensurable_conj {H K : subgroup G} (g : G) :
+   commensurable H K ↔ commensurable (H.conj_subgroup g) (K.conj_subgroup g) :=
+and_congr (not_iff_not.mpr (eq.congr_left (cardinal.to_nat_congr (quot_conj_equiv H K g))))
+  (not_iff_not.mpr (eq.congr_left (cardinal.to_nat_congr (quot_conj_equiv K H g))))
+
+lemma commensurable_inv (H : subgroup G) (g : G) :
+  commensurable (subgroup.conj_subgroup g H) H ↔ commensurable H (subgroup.conj_subgroup g⁻¹ H)  :=
+begin
+  rw [commensurable_conj g⁻¹ ,← subgroup.conj_subgroup_mul g⁻¹ g],
+  simp only [mul_left_inv, subgroup.cong_subgroup_one],
+end
+
+/--For `H` a subgroup of `G`, this is the subgroup of all elements `g : G`
+such that `commensurable (conj_subgroup g H) H`   -/
+
+def commensurator (H : subgroup G) : subgroup G :=
+{ carrier := {g : G | commensurable (subgroup.conj_subgroup g H) H },
+  one_mem' := by {simp only [set.mem_set_of_eq, subgroup.cong_subgroup_one], },
+  mul_mem' := by {intros a b ha hb,
+      simp only [set.mem_set_of_eq] at *,
+      rw subgroup.conj_subgroup_mul,
+      have h1 : commensurable (subgroup.conj_subgroup a (subgroup.conj_subgroup b H))
+        (subgroup.conj_subgroup a H),
+      { have hab := trans ha ((comm ).1 hb),
+        rw (commensurable_conj a⁻¹) at hab,
+        rw ← subgroup.conj_subgroup_mul at hab,
+        simp only [mul_left_inv, subgroup.cong_subgroup_one] at hab,
+        have r1 := commensurable_inv (subgroup.conj_subgroup b H) a,
+        have hab2 := trans hb hab,
+        have r2 := r1.2 hab2,
+        apply trans r2 (trans hb (symm  ha)),},
+      exact trans h1 ha,},
+  inv_mem' := by {simp only [set.mem_set_of_eq],
+      intros x hx,
+      rw comm,
+      apply (commensurable_inv H x).1 hx,},}
```

Comment:
If you do this, then you could define `commensurator` by using the isomorphism between `G` and `conj_act G`.
```suggestion
def commensurator' (H : subgroup G) : subgroup (conj_act G) :=
{ carrier := {g : conj_act G | commensurable (g • H) H},
  one_mem' := by rw [set.mem_set_of_eq, one_smul],
  mul_mem' := λ a b ha hb, by
  { rw [set.mem_set_of_eq, mul_smul],
    exact trans ((commensurable_conj a).mp hb) ha },
  inv_mem' := λ a ha, by rwa [set.mem_set_of_eq, comm, ←commensurable_inv] }
```

Reply:
So I would prefer to have the commensurator to be a subgroup of `G` instead of `conj_act G`. So I used this definition an the map `conj_act G ->* G` to define it as the image. The proof that it gives the "right" thing is a bit ugly, so maybe you can see an nicer way to prove it :)